docs/java.md

```diff
@@ -0,0 +1,373 @@
+
+## Java Documentation
+
+[The complete Java example is here](SpellExample.java)
+
+#### How to parse query results
+
+
+Every fauna query returns a `ListenableFuture<Value>`.  Value is the json representation of the query response.  It can be pretty printed by adding a jackson dependency:
```

Comment:
We've been trying to move away from `curl` as well as the notion that we use JSON over the wire. Specially because at some point that will be replaced by a binary protocol.

I understand that JSON is human readable, hence, the desire to see the Fauna values as JSON objects. However, I feel like that is confusing, specially when people stumble upon non-standard JSON values. E.g.: Ref, Date, Time, and Query.

IMO, we should treat values as opaque objects until converted to a native type. Nevertheless, we also need to work on the string representation for all values so that `System.out.println(..)` gives people useful information.

I believe the docs are currently just showing each value's `toString`. Perhaps we should do the same so we don't lead people to see Fauna values as JSON. That vision, once established, is hard to get around.

I'll let @klavavej give the final word about it, though.

Reply:
I can understand not wanting users to think of Fauna values as JSON and the possible confusion that specialized types can cause.   The challenge is looking at the JSON output is currently the only way I know do debug problems working with the return results and understanding how to get the right results out of fauna.

Reply:
Value.toString should be useful output. If that's not the case then we should fix that.

Reply:
I think for Value.toString to be useful it would need to do some basic formatting of the output.  Something like nested indenting and splitting each element in an array onto a separate line?  Right now it is just a giant output string that I have a hard time parsing without a pretty printer.

Reply:
be aware that changing toString right now would make @jfmiii very upset as he worked on java examples in docs to get the output of the examples 😃

Reply:
@jfmiii - Do tell what is your secret?  How do the docs get the nice output formatting from the toString?  Because what I see the docs doesn't look anything like what I get when I run the examples.  For example when I drop this line from the docs:

```java
   System.out.println(
            client.query(
                Paginate(
                    Match(Index(Value("spells_by_element")), Value("fire")))
            ).get());
```

I get just a plain string:

`ObjectV({data=Arr([RefV(id = "201833952983384576", class = RefV(id = "spells", class = RefV(id = "classes"))), RefV(id = "201833952983386624", class = RefV(id = "spells", class = RefV(id = "classes")))])})`

And not the nicely formatted string I see in the docs?

Reply:
From what I’ve seen, the Java responses currently always give human readable information from System.out.println(..) without an extra step of converting to JSON. Though I’ve certainly seen output from Javascript and Go that aren’t human readable. 

For example the Query API reference - https://fauna.com/documentation/reference/queryapi
the sample under “with sources:”
---
Java output is fully human readable
=> ObjectV({
 data=Arr([
   ObjectV({
     value=RefV(id = “181388642046968320”, class = RefV(id = “spells”, class = RefV(id = “classes”))),
     sources=Arr([
       SetRefV({
          match=RefV(id = “spells_by_element”, class = RefV(id = “indexes”)),
          terms=StringV(fire)
       })
     ])
   }),
   ObjectV({
     value=RefV(id = “181388642071085568", class = RefV(id = “spells”, class = RefV(id = “classes”))),
     sources=Arr([
       SetRefV({
          match=RefV(id = “spells_by_element”, class = RefV(id = “indexes”)),
          terms=StringV(fire)
       }),
       SetRefV({
               match=RefV(id = “spells_by_element”, class = RefV(id = “indexes”)),
               terms=StringV(water)
       })
     ])
   }),
   ObjectV({
     value=RefV(id = “181388642088911360”, class = RefV(id = “spells”, class = RefV(id = “classes”))),
     sources=Arr([
       SetRefV({
          match=RefV(id = “spells_by_element”, class = RefV(id = “indexes”)),
          terms=StringV(water)
       })
     ])
   })
 ])
})
---
whereas javascript obscures sources as “object”
=> { data:
  [ { value: Ref(id=181388642046968320, class=Ref(id=spells, class=Ref(id=classes))),
      sources: [Object] },
    { value: Ref(id=181388642071085568, class=Ref(id=spells, class=Ref(id=classes))),
      sources: [Object] },
    { value: Ref(id=181388642088911360, class=Ref(id=spells, class=Ref(id=classes))),
      sources: [Object] } ] }
---
And Go obscures the class
=> map[data:[
 map[value:{181388642046968320 0xc42024a960 <nil>} sources:[{map[match:{spells_by_element 0xc42024ab00 <nil>} terms:fire]}]]
 map[value:{181388642071085568 0xc42024ada0 <nil>} sources:[{map[terms:fire match:{spells_by_element 0xc42024af40 <nil>}]} {map[match:{spells_by_element 0xc42024b100 <nil>} terms:water]}]]
 map[value:{181388642088911360 0xc42024b3e0 <nil>} sources:[{map[match:{spells_by_element 0xc42024b580 <nil>} terms:water]}]]
]]
---
John manually inserted line breaks to make the responses more readable in the docs but as far as I know he didn't add any extra code to convert obscured info to human readable info.

Reply:
My opinion was that manually inserted line breaks are a reasonable deviation from the actual output for the purpose of the docs given the maximum width of response viewport in the docs and the lack of wrapping therein. 

I thought the linebreaks in the responses docs would increase usability of the docs enough to outweigh any potential confusion about the actual response not having matching linebreaks

Reply:
So, what is the decision here?

I feel from the conversation that we're leaning towards removing te JSON output so that we don't induce people to view Fauna values as JSON values.

Is that correct?

Reply:
@erickpintor - I'm not sure I understand the options you're presenting. When you suggest "removing te JSON output so that we don't induce people to view Fauna values as JSON values." are you suggesting that we don't include output of any kind? or are you suggesting that we replace it with a different kind of output?

FWIW, the docs currently explicitly mention JSON in a few places. 

https://fauna.com/documentation/howto/crud
- in the Java Prerequisites "Jackson for JSON parsing."

https://fauna.com/documentation/reference/queryapi
- "Booleans, Integers, Decimals, Strings, and Null JSON literals all evaluate to themselves:"
- "Because regular JSON objects represent function calls, object values must be constructed using object:"
- "FaunaDB supports some types beyond those native to JSON. Special types are encoded as JSON objects with a single field."

https://fauna.com/documentation/reference/glossary
- in the table for Classes "data" : "A JSON object. Optional."
- in the table for Databases "data" : "A JSON object. Optional."
- in the table for Keys "data" : "A JSON object. Optional."
- "Instances returned by queries are represented as JSON objects. "

I raise these current mentions of JSON just to say that not encouraging people to view Fauna Values as JSON values may be a wider change than just the scope of the output in the JavaDocs

Reply:
@klavavej: This code has a function that displays Fauna values as JSON rather than the standard Java output, like the online docs. When I said to remove the JSON output, I meant to remove that function so that both this code and the docs are presenting the same output to the users, even if not pretty printed.

I do agree that removing JSON from our shared vocabulary is a big change in which I am not really in a position to make a decision. However, I had the impression that it was the path we wanted to move towards. If that is correct, we should not introduce more code that refers to Fauna values as JSON so that we can gradually stop using JSON related terms across our ecosystem.

@retroryan I understand and I do agree that, right now, the JSON output that you have is much more readable than current implementation of `Value.toString`. What I wanted to make sure is that we make a conscious decision in favor or against this patter. I imagine it looks simple since most developers are familiar with JSON but, the accidental effects of trying to introspect into the values have gotten people stuck before.

Reply:
Seems like I accidentally stirred a much bigger conversation than I expected :)

For this PR what do we want to do?

Reply:
Thanks @erickpintor! I am in favor of the output here being standard Java output to match the docs side

Reply:
The only challenge is the docs site is not the standard Java output.  The standard doc output has been hand formatted to be readable.  I am not sure of the process but someone has been formatting the doc site to add line breaks and indents to the output.  The standard java output looks like this:

```
Stored spell:
 ObjectV({ref=RefV(id = "201933794520859136", class = RefV(id = "spells", class = RefV(id = "classes"))), ts=LongV(1528837923444036), data=ObjectV({cost=LongV(25), name=StringV(Water Dragon's Claw), element=StringV(water)})})
```

And if we add output formatting we get this instead:

```
Stored spell:
 {
  "object" : {
    "ref" : {
      "@ref" : {
        "id" : "201933878650208768",
        "class" : {
          "@ref" : {
            "id" : "spells",
            "class" : {
              "@ref" : {
                "id" : "classes"
              }
            }
          }
        }
      }
    },
    "ts" : 1528838003677989,
    "data" : {
      "object" : {
        "cost" : 25,
        "name" : "Water Dragon's Claw",
        "element" : "water"
      }
    }
  }
}
```

Reply:
I'll throw a random idea here:

What if we don't print the raw values at all? Instead, in places where the data is relevant, we extract and convert the relevant information before presenting it.

That way we reinforce that Fauna values should be treated as opaque values until converted to a native Java type.

For example, in the output above, the verboseness of the recursive reference makes it difficult to understand. Instead, if we want to teach about instances' references, we could do something like:

```java
Value ref = createResult.at("ref");

//...

Value readResult = client.query(Get(ref)).get();

String name = readResult.at("data", "name").get(String.class).get();
Integer cost = readResult.at("data", "cost").get(Integer.class).get();
String element = readResult.at("data", "element").get(String.class).get();
```

For the purpose of understating that instances are referenceable, how references are printed out in the console doesn't matter that much.

Users will inevitably print out our values while learning how yo user our system. That's why we need to fix the `Value.toString`. However, as we walk them through how to use our driver, we can focus more on the aspects of the driver usage, like expression composition, data extraction, and data conversion.

Reply:
@erickpintor - I removed the pretty printing and changed everything just to print out the raw output.  Then in a couple of cases I showed as you suggested, how to do data extraction and data conversion.  Let me know if any more changes are needed.